File: src/main/java/me/kzv/ecommerce/security/CustomUserDetailsService.java
package me.kzv.ecommerce.security;

import lombok.RequiredArgsConstructor;
import me.kzv.ecommerce.domain.member.Member;
import me.kzv.ecommerce.domain.member.SocialMember;
import me.kzv.ecommerce.domain.member.MemberRepository;
import org.springframework.security.core.userdetails.UserDetails;
import org.springframework.security.core.userdetails.UserDetailsService;
import org.springframework.security.core.userdetails.UsernameNotFoundException;
import org.springframework.stereotype.Service;

@Service
@RequiredArgsConstructor
public class CustomUserDetailsService implements UserDetailsService {

    private final MemberRepository memberRepository;

    @Override
    public UserDetails loadUserByUsername(String username) throws UsernameNotFoundException {
        Member member = memberRepository.findByEmail(username).orElseThrow(() -> {
            throw new UsernameNotFoundException("이메일 또는 비밀번호를 확인해주세요");
        });

        var isSocial = (SocialMember) member;
        if (isSocial.getSocialType() != null) { // 로컬 회원이 아닌 소셜로 가입한 회원
            throw new UsernameNotFoundException("이메일 또는 비밀번호를 확인해주세요");
        }


        return MemberSecurityContext.of(member);
    }

}

File: src/main/kotlin/me/kzv/crud/security/service/CustomUserDetailsService.kt
package me.kzv.crud.security.service

import me.kzv.crud.repository.MemberRepository
import org.springframework.data.repository.findByIdOrNull
import org.springframework.security.core.GrantedAuthority
import org.springframework.security.core.authority.SimpleGrantedAuthority
import org.springframework.security.core.userdetails.UserDetails
import org.springframework.security.core.userdetails.UserDetailsService
import org.springframework.security.core.userdetails.UsernameNotFoundException
import org.springframework.stereotype.Service


@Service
class CustomUserDetailsService(
    private val memberRepository: MemberRepository,
) : UserDetailsService {

    override fun loadUserByUsername(email: String): UserDetails {
        val member = memberRepository.findByIdOrNull(email) ?: throw UsernameNotFoundException("유저가 존재하지 않습니다.")

        val grantedAuthority: MutableList<GrantedAuthority> = arrayListOf()
        member.authorities.forEach { authorityType -> grantedAuthority.add(SimpleGrantedAuthority(authorityType.toString())) }

        return MemberContext(member, grantedAuthority)
    }
}